ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java

```diff
@@ -60,6 +63,9 @@
   // Resources that need to be shut down when the router does.
   private final List<Closeable> resourcesToClose;
   private final AtomicInteger currentBackgroundOperationsCount = new AtomicInteger(0);
+  private static final int cacheMaxLimit = 1000;
+  // Cache to store blob IDs which were not found in servers recently.
+  private final Cache<String, Boolean> notFoundcache;
```

Comment:
what is the motivation behind this cache? Are we seeing same blob ids are retried multiple times?

Reply:
Negative cache was suggested by @tbingol.  Since we are making more strict checks for terminating the operations based on local not-found responses, we might have more cross colo look ups. Using not-found or negative cache might help us reduce the load on servers if clients were to do immediate look-ups again.